_articles/aws_sts_storage_locations.md

```diff
@@ -0,0 +1,178 @@
+---
+title: "AWS Security Token Service Storage Locations"
+layout: article
+excerpt: Follow these steps to set up STS Storage Locations and use them to access your S3 data directly.
+category: admin-and-settings
+---
+
+Using AWS Security Token Service (STS), Synapse can securely grant you temporary AWS credentials with access to data directly in S3. This can be useful if you want to:
+
+* Transfer data in bulk
+* Allow your compute cluster to read S3 objects using the S3 APIs
+
+All of which you can now do with minimal overhead from Synapse.
+
+{% include note.html content="You can only create STS Storage Locations on an empty folder. This is to ensure consistency between the STS Storage Location and the Synapse folder." %}
+
+## Synapse-Managed STS Storage Locations
+
+You can create an STS Storage Location using Synapse storage. Synapse will store the files for you, but you can still get temporary S3 credentials, scoped to just the files and folders within your STS Storage Location.
+
+To set up the STS Storage Location using Synapse storage, first make sure you have an empty Synapse folder, then run the following code:
+
+##### Python
+
+```python
+# Set storage location
+import synapseclient
+import json
+syn = synapseclient.login()
+FOLDER = 'syn12345'
+
+destination = {'uploadType':'S3',
+               'stsEnabled':True,
+               'concreteType':'org.sagebionetworks.repo.model.project.S3StorageLocationSetting'}
+destination = syn.restPOST('/storageLocation', body=json.dumps(destination))
+
+project_destination ={'concreteType': 'org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
+                      'settingsType': 'upload'}
+project_destination['locations'] = [destination['storageLocationId']]
+project_destination['projectId'] = FOLDER
+
+project_destination = syn.restPOST('/projectSettings', body = json.dumps(project_destination))
+```
+
+##### R
+
+```r
+#set storage location
+library(synapser)
+library(rjson)
+synLogin()
+folderId <- 'syn12345'
+
+destination <- list(uploadType='S3',
+                    stsEnabled=TRUE,
+                    concreteType='org.sagebionetworks.repo.model.project.S3StorageLocationSetting')
+destination <- synRestPOST('/storageLocation', body=toJSON(destination))
+
+projectDestination <- list(concreteType='org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
+                           settingsType='upload')
+projectDestination$locations <- list(destination$storageLocationId)
+projectDestination$projectId <- folderId
+
+projectDestination <- synRestPOST('/projectSettings', body=toJSON(projectDestination))
+```
+
+You can then upload files to this STS Storage Location using the normal methods, either through [the Synapse website](https://www.synapse.org/) or through the client of your choice.
```

Comment:
I see how it is possible to add FileHandles for files uploaded via STS tokens (or otherwise) to an external bucket using the **POST /externalFileHandle/s3** endpoint. However for Synapse storage (i.e. non-external), after getting an STS token and using it to upload a file directly to a Synapse managed S3 bucket I don't see an equivalent way to be able to add a FileHandle representation for that uploaded file. The externalFileHandle/s3 endpoint doesn't work properly rejecting it as a non external storage location.

Also, per @thomasyu888 's point about the accessibility of the bucket name, to upload a file at all to Synapse storage using an STS token and an external S3 client one must already know the bucket name as it doesn't seem to be accessible via the upload destination or storage location.

Reply:
Good catch. I've filed https://sagebionetworks.jira.com/browse/PLFM-6227 to address this.